### PR TITLE
Flattens The Floor Plane (Camera Update Too)

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Ok so like, side map right? It makes things higher up in the world render above things lower down in the world.

Most of the time this is what we want, but it is NOT what we want for floors.
Floors are allowed to be larger then 32x32, and if they are we want them to render based off JUST their layer.
If we don't allow this grass turfs and others get cut off on their bottom edge, which looks WEIRD.

In order to make this happen, we can add TOPDOWN_LAYER to every layer on the floor plane and disable sidemap.

I've added documentation for this to VISUALS.md, and have also implemented unit test errors to prevent mixing TOPDOWN layers with non topdown planes (or vis versa).
This new test adds ~1 second to tests, which is I think a perfectly scrumpulent number.

EDIT:

I nerd sniped myself and implemented sidemap layering and lighting for cameras (also larger then 32x32 icon support for getflat)
The lighting isn't perfect, we don't handle things displaying in the void all that well (I am convinced getflat blending is broken but I have no debugger so I can't fix it properly), but it'll do.

This came up cause I had to fix another layering issue in cameras and thought I might as well go all in.
image

## Why It's Good For The Game

Old:
image

New:
image

## Changelog
:cl:
fix: Grass turfs will render properly now. Reworked how floors render, please report any bugs!
fix: Cameras now properly capture lighting
fix: The layering seen in photos should better match the actual game
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
does this also fix cult rune rendering or was that fixed a while ago? regardless this is a good change and a reminder that the byond renderer must be absolutely fucking cursed

**Comment by EEASAS** [comment]
i kinda liked the lime coloured grass but oh well

**Comment by LemonInTheDark** [comment]
> i kinda liked the lime coloured grass but oh well

to be clear I used the greener stuff basically on accident, the color of the grass isn't being changed at all it's just the grass tufts actually show on the south edge now

**Comment by LemonInTheDark** [comment]
> does this also fix cult rune rendering or was that fixed a while ago? regardless this is a good change and a reminder that the byond renderer must be absolutely fucking cursed

I do not know. frankly right now it breaks vents displaying because I forgot to test for broken combos AFTER init, but that is LIFE baby

**Comment by LemonInTheDark** [comment]
> does this also fix cult rune rendering or was that fixed a while ago? regardless this is a good change and a reminder that the byond renderer must be absolutely fucking cursed

Likely no because iirc it's not on the floor plane at all

Also this should be good now I fixed the rest of things + a unit test

File: code/modules/unit_tests/plane_sanity.dm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/// Ensures we have no invalid plane/layer combos post init
+/datum/unit_test/plane_layer_sanity
+
+/datum/unit_test/plane_layer_sanity/Run()
+	// This fucker's gonna be slow, I'm sorry
```
Comment (san7890) [inline code/modules/unit_tests/plane_sanity.dm:5]: add `priority = TEST_LONGER`?

Comment (LemonInTheDark) [reply code/modules/unit_tests/plane_sanity.dm:6]: sure

**Comment by san7890** [comment]
test merges going alright?

**Comment by LemonInTheDark** [comment]
https://github.com/tgstation/tgstation/issues/84524
no :(

**Comment by PapaMichae1** [comment]
I don't know the specifics well enough to attempt a issue yet but cameras seem broken in this tm too

**Comment by jlsnow301** [comment]
20 rounds on the tm + maintainer pr + lgtm

**Comment by LemonInTheDark** [comment]
> ... maintainer pr ...

you should not trust me

**Comment by OverwatchVoice** [comment]
As a result of this PR, the atmos-thermal vision is still broken, making it even more difficult to locate breaches on multi-Z maps.

**Pull request merged**